src/learn/database/index.md

```diff
@@ -0,0 +1,688 @@
+
+{% endbox %}
+
+### Formatting columns
+
+Our `created_at` column is a full timestamp, with a date and time, which is not particularly human-readable. Let's imagine our app is only for tracking tasks on the same day, so we only care about the time part. We can amend our query to use the `TIME`
+
+```js
+const select_tasks = db.prepare(/*sql*/ `
+  SELECT
+    id,
+    content,
+    TIME(created_at)
+  FROM tasks
+`);
+```
+
+If you log the result of this query you should see the task objects change:
+
+```json
+[{ "id": 2, "content": "Send mum flowers", "TIME(created_at)": "08:52:30" }]
+```
+
+However the column name has changed to represent this. Ideally we would keep the same name (`created_at`). We can rename columns using the `AS` SQL operator:
+
+```js
+const select_tasks = db.prepare(/*sql*/ `
+  SELECT
+    id,
+    content,
+    TIME(created_at) AS created_at
+  FROM tasks
+`);
+```
+
+If you run this again you should see the object key is now `created_at` like before.
+
+```json
+[{ "id": 2, "content": "Send mum flowers", "created_at": "08:52:30" }]
+```
+
+## Deleting a row
+
+Let's add another model function that can delete a task from our `tasks` table. It should take an ID parameter and delete just the row that matches.
+
+```js
+const delete_task = db.prepare(/*sql*/ `
+  DELETE FROM tasks WHERE id = ?
+`);
+
+function removeTask(id) {
+  return delete_task.run(id);
+}
+```
+
+{% box "error" %}
+
+SQL queries act on **every row** by default. This makes deletion and updates very dangerous. Without a `WHERE` clause or other condition they will delete/update _every_ row.
+
+{% endbox %}
+
+Test this by calling `removeTasks(1)` to delete the first task you created earlier. You can call `listTasks` to check that a task has been removed.
+
+## Testing the model
+
+So far we've been manually testing things by calling our model functions. It would be better to write some automated tests to make this reproducible and reusable. That way we can catch reversions if the code breaks later on.
+
+Let's write a test to verify several of our model functions. Create a new file `test/tasks.test.js`:
+
+```js
+const test = require("node:test");
+const assert = require("node:assert");
+const model = require("../model/tasks.js");
+const db = require("../database/db.js");
+
+test("can create, remove & list tasks", () => {
+  db.exec("DELETE FROM tasks");
+
+  const task = model.createTask("test task");
+  assert.equal(task.id, 1);
+  assert.equal(task.content, "test task");
+  assert.equal(task.complete, 0);
+
+  model.removeTask(task.id);
+  const tasks = model.listTasks();
+  assert.equal(tasks.length, 0);
+});
+```
+
+We can run the test directly with:
+
+```shell
+node test/tasks.test.js
+```
```

Comment:
Running `node test/tasks.test.js` gave me an error message:

```
failureType: 'testCodeFailure'
  error: 'Too many parameter values were provided'
  code: 'ERR_TEST_FAILURE'
  stack: |-
    Object.createTask (/Users/nkeemer/Documents/FACReviewNew/Week2/learn-database/model/tasks.js:26:24)
    TestContext.<anonymous> (/Users/nkeemer/Documents/FACReviewNew/Week2/learn-database/test/tasks.test.js:9:22)
```

COMMENTS: This seemed to be resolved by making the argument for  createTask an object, e.g.
Changing ....
`const task = model.createTask("test task"); `
 To ...
` const task = model.createTask({content: "test task", complete: 0});`


Running `node test/tasks.test.js` again gave me another error message, caused by `assert.equal(task.id, 1);`:

```
failureType: 'testCodeFailure'
  error: '2 == 1'
  code: 'ERR_ASSERTION'
```

COMMENTS: It seems Model.js is being created first, therefore making “test task” id = 2, not 1 as asserted?